### PR TITLE
Extract all Ecto functionality to separate module

### PR DESCRIPTION
**Pull request by jsteiner**
This should make it easier to add other adapters in the future and
reduces the need for conditionals due to separation of concerns.

### CODE REVIEW

File: lib/ex_machina/ecto.ex [hunk 1]
```diff
@@ -0,0 +1,58 @@
+defmodule ExMachina.Ecto do
+  defmacro __using__(opts) do
+    quote do
+      use ExMachina
+
+      @repo Dict.get(unquote(opts), :repo)
```
Comment (jsteiner) [inline lib/ex_machina/ecto.ex]: What is the best way to make this option required?

File: lib/ex_machina.ex [hunk 2]
```diff
@@ -20,16 +20,15 @@ defmodule ExMachina do
 
   defmodule UndefinedSave do
     @moduledoc """
-    Error raised when trying to call create and no repo or save_function is
-    defined.
+    Error raised when trying to call create and save_record/1 is
+    not defined.
```
Comment (paulcsmith) [inline lib/ex_machina.ex]: Heh, nice catch :)

File: lib/ex_machina/ecto.ex (hunk 1 above)
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: Dict.fetch! would blow up if you try to call it without repo defined in the opts. However if you want something more friendly you could probably do something in save record like we did before. So you would check if `@repo` is defined. If it is, save it, if not raise a helpful error letting them know they need to pass `repo` in to the opts

File: lib/ex_machina.ex [hunk 3]
```diff
@@ -279,10 +194,41 @@ defmodule ExMachina do
...
+      If using ExMachina.Ecto (`use ExMachina.Ecto, repo: MyApp.Repo`) this
+      function will call `insert!` on the passed in repo.
+
+      If you are not using ExMachina.Ecto, you must define a custom
+      save_record/1 for saving the record.
+
+      ## Examples
+
+          defmodule MyApp.Factories do
+            use ExMachina, repo: MyApp.Repo
```
Comment (paulcsmith) [inline lib/ex_machina.ex]: use ExMachina.Ecto

File: lib/ex_machina.ex [hunk 4]
```diff
@@ -279,10 +194,41 @@ defmodule ExMachina do
...
+
+            def factory(:user), do: %User{name: "John"}
+          end
+
+          # Will build and save the record to the MyApp.Repo
+          MyApp.Factories.create(:user)
+
+          defmodule MyApp.JsonFactories do
+            # Note `repo` was not passed as an option
```
Comment (paulcsmith) [inline lib/ex_machina.ex]: Note we are using just `ExMachina` without Ecto and without the `repo` option?

File: lib/ex_machina/ecto.ex (hunk 1 above)
Comment (jsteiner) [inline lib/ex_machina/ecto.ex]: Is there any way to get it to crash at compile time but also provide a friendly message? Maybe `Dict.fetch!` is good enough?

File: lib/ex_machina/ecto.ex [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  defp drop_ecto_fields(record = %{__struct__: struct, __meta__: %{__struct__: Ecto.Schema.Metadata}}) do
+    record
+    |> Map.from_struct
+    |> Map.delete(:__meta__)
+    |> Map.drop(struct.__schema__(:associations))
+  end
+  defp drop_ecto_fields(record) do
+    raise ArgumentError, "#{inspect record} is not an Ecto model. Use `build` instead."
+  end
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: I'm wondering if `assoc` should be only in Ecto and should return the id automatically. It's only useful for Ecto, because you can just use `build(:association)` for other types of factories. What do you think?

File: lib/ex_machina/ecto.ex (hunk 1 above)
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: If you use Dict.fetch! It should crash at compile time, but I don't know of a way to get a nice error message. You could try get the option before the `quote` and raising there, but I've never done that

File: test/ex_machina/ecto_test.ex [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  test "create_list/3 creates factory and saves it passed in number of times" do
+    users = MyApp.ExMachina.create_list(3, :user, admin: true)
+
+    created_user = %{id: 3, name: "John Doe", admin: true}
+    assert users == [created_user, created_user, created_user]
+    assert_received {:created, ^created_user}
+    assert_received {:created, ^created_user}
+    assert_received {:created, ^created_user}
+    refute_received {:created, _}
+  end
```
Comment (paulcsmith) [inline test/ex_machina/ecto_test.ex]: I think we could remove all of the `create...` tests and instead call `save_record` and ensure that it saves. This would simplify this test file a lot I think. 

create, create_list, etc. are tested in the other file already

File: test/ex_machina_test.exs [hunk 7]
```diff
@@ -171,14 +131,15 @@ defmodule ExMachinaTest do
     user = MyApp.ExMachina.create(:user, admin: true)
 
     assert user == %{id: 3, name: "John Doe", admin: true}
-    assert_received {:created, %{name: "John Doe", admin: true}}
+    assert_received {:custom_save, %{name: "John Doe", admin: true}}
   end
 
   test "create/2 builds factory and performs custom save if repo is not set" do
```
Comment (paulcsmith) [inline test/ex_machina_test.exs]: "...and performs save with user defined `save_record/1`"

File: lib/ex_machina/ecto.ex (hunk 5 above)
Comment (jsteiner) [inline lib/ex_machina/ecto.ex]: Sounds good. We can always move it back later if we have to.

File: test/ex_machina_test.exs [hunk 8]
```diff
@@ -171,14 +131,15 @@ defmodule ExMachinaTest do
     user = MyApp.ExMachina.create(:user, admin: true)
 
     assert user == %{id: 3, name: "John Doe", admin: true}
-    assert_received {:created, %{name: "John Doe", admin: true}}
```
Comment (paulcsmith) [inline test/ex_machina_test.exs]: We can probably remove this test and just use the one below since we aren't doing the repo definition here anymore

File: test/ex_machina_test.exs [hunk 9]
```diff
@@ -1,30 +1,9 @@
...
-  defmodule MyApp.Book do
-    defstruct title: nil, publisher: nil, __meta__: %{__struct__: Ecto.Schema.Metadata}, publisher_id: 1
-
-    def __schema__(:associations) do
-      [:publisher]
-    end
-  end
-
   defmodule MyApp.ExMachina do
     use ExMachina, repo: TestRepo
```
Comment (paulcsmith) [inline test/ex_machina_test.exs]: I think this should not have the `repo` option right?

**Comment by paulcsmith** [comment]
Left a couple comments. This look really awesome! I think this was a great idea. Also, we'll need to update the README

File: lib/ex_machina/ecto.ex [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+  @doc """
+  Saves a record using `Repo.insert!` when `create` is called.
+  """
+  def save_record(module, repo, record) do
+    if repo do
+      repo.insert!(record)
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: I think the simplest thing to do would be to raise a helpful error here if the repo is nil

File: test/ex_machina_test.exs [hunk 11]
```diff
@@ -171,14 +131,15 @@ defmodule ExMachinaTest do
...
+    record = MyApp.ExMachina.create(:user)
 
-    assert record == %{foo: :bar}
-    assert_received {:custom_save, %{foo: :bar}}
+    created_record = %{admin: false, id: 3, name: "John Doe"}
+    assert record == created_record
+    assert_received {:custom_save, ^created_record}
   end
 
   test "create/2 raises a helpful error if no repo and no custom function" do
```
Comment (paulcsmith) [inline test/ex_machina_test.exs]: replace "if no repo and no custom function" with "if no custom save_record/1 function is defined" or something like that?

File: lib/ex_machina/ecto.ex [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+      use ExMachina
+
+      @repo Dict.fetch!(unquote(opts), :repo)
+
+      def fields_for(factory_name, attrs \\ %{}) do
+        ExMachina.Ecto.fields_for(__MODULE__, factory_name, attrs)
+      end
+
+      defp assoc(attrs, factory_name, opts \\ []) do
+        ExMachina.assoc(__MODULE__, attrs, factory_name, opts)
```
Comment (paulcsmith) [inline lib/ex_machina/ecto.ex]: Does this work? I think it's defined on `ExMachina.Ecto`

File: test/ex_machina/ecto_test.ex [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        id: 1,
+        title: "My Awesome Article",
+        author_id: assoc(attrs, :author, factory: :user).id
+      }
+    end
+
+    def factory(:comment, attrs) do
+      %{
+        body: "This is great!",
+        article_id: assoc(attrs, :article).id
```
Comment (paulcsmith) [inline test/ex_machina/ecto_test.ex]: Did we want to make this return id by default, or leave it as-is?

File: test/ex_machina/ecto_test.ex [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+  defmodule MyApp.Book do
+    defstruct title: nil, publisher: nil, __meta__: %{__struct__: Ecto.Schema.Metadata}, publisher_id: 1
+
+    def __schema__(:associations) do
+      [:publisher]
+    end
+  end
+
+  defmodule MyApp.ExMachina do
```
Comment (paulcsmith) [inline test/ex_machina/ecto_test.ex]: What do you think of calling this just `EctoFactories`?

File: test/ex_machina/ecto_test.ex (hunk 13 above)
Comment (jsteiner) [inline test/ex_machina/ecto_test.ex]: I think we can leave it as is for now. Now that we've extracted it to it's own module, we can make another commit to allow for `article: assoc(attrs, :article)`

Comment (paulcsmith) [inline test/ex_machina/ecto_test.ex]: Ah ok, good idea!

**Comment by paulcsmith** [comment]
:+1: looks great!
